server/src/test/java/org/elasticsearch/transport/RemoteClusterAwareClientTests.java

```diff
@@ -62,6 +72,95 @@ private MockTransportService startTransport(String id, List<DiscoveryNode> known
+                // the TaskCancellationService references the same TransportService instance
+                // this is identically to how it works in the Node constructor
+                localService.getTaskManager().setTaskCancellationService(new TaskCancellationService(localService));
+                localService.start();
+                localService.acceptIncomingRequests();
+
+                SearchShardsRequest searchShardsRequest = new SearchShardsRequest(
+                    new String[] { "test-index" },
+                    IndicesOptions.strictExpandOpen(),
+                    new MatchAllQueryBuilder(),
+                    null,
+                    "index_not_found", // this request must fail
+                    randomBoolean(),
+                    null
+                );
+                Task parentTask = localService.getTaskManager().register("test_type", "test_action", searchShardsRequest);
+                TaskId parentTaskId = new TaskId("test-mock-node-id", parentTask.getId());
+                searchShardsRequest.setParentTask(parentTaskId);
+                var client = new RemoteClusterAwareClient(
+                    localService,
+                    "cluster1",
+                    threadPool.executor(TEST_THREAD_POOL_NAME),
+                    randomBoolean()
+                );
+
+                AtomicBoolean cancelChildReceived = new AtomicBoolean(false);
+                remoteTransport.addRequestHandlingBehavior(
+                    TaskCancellationService.CANCEL_CHILD_ACTION_NAME,
+                    (handler, request, channel, task) -> {
+                        handler.messageReceived(request, channel, task);
+                        cancelChildReceived.set(true);
+                    }
+                );
+                AtomicLong searchShardsRequestId = new AtomicLong(-1);
+                AtomicBoolean cancelChildSent = new AtomicBoolean(false);
+                localService.addSendBehavior(remoteTransport, (connection, requestId, action, request, options) -> {
+                    connection.sendRequest(requestId, action, request, options);
+                    if (action.equals("indices:admin/search/search_shards")) {
+                        searchShardsRequestId.set(requestId);
+                    } else if (action.equals(TaskCancellationService.CANCEL_CHILD_ACTION_NAME)) {
+                        cancelChildSent.set(true);
+                    }
+                });
+
+                // assert original request failed
+                var future = new PlainActionFuture<SearchShardsResponse>();
+                client.execute(TransportSearchShardsAction.REMOTE_TYPE, searchShardsRequest, future);
+                ExecutionException e = expectThrows(ExecutionException.class, future::get);
+                assertThat(e.getCause(), instanceOf(RemoteTransportException.class));
+
+                // assert remote task is cancelled
+                assertBusy(cancelChildSent::get);
```

Comment:
This was not asserting true but just calling the .get().

```suggestion
                assertBusy(() -> {
                    assertTrue(cancelChildSent.get());
                    assertTrue(cancelChildReceived.get());
                });
```

And remove the `verify` assertion and the `spy` on the task manager.

Reply:
Good catch with `assertBusy`! I fixed it with `CountDownLatch`.

> I think I'm a bit skeptical now about using the spy. spy() introduces a memory overhead as mockito then remembers all invocations, and that's why we probably saw the OOM in CI. And maybe we should try to avoid it if we can. I think if we correct a thing in the test (see my other comment), we can remove the spy altogether.

I feel you! I've pushed  https://github.com/elastic/elasticsearch/pull/109440/commits/3e0a6352fa4e2e3acb93215bd1d2edbaa661396f to add a new test setting to conditionally wrap the task manager with `spy`. That's because I think I still want to keep the `verify` call, because I can also check the parent task id.

Please take another look.